lib/internal/modules/esm/translators.js

```diff
@@ -219,38 +317,43 @@ function cjsPreparseModuleExports(filename) {
   const exportNames = new SafeSet(new SafeArrayIterator(exports));
 
   // Set first for cycles.
-  cjsParseCache.set(module, { source, exportNames, loaded });
+  cjsParseCache.set(module, { source, exportNames });
 
   if (reexports.length) {
     module.filename = filename;
     module.paths = CJSModule._nodeModulePaths(module.path);
-  }
-  ArrayPrototypeForEach(reexports, (reexport) => {
-    let resolved;
-    try {
-      resolved = CJSModule._resolveFilename(reexport, module);
-    } catch {
-      return;
-    }
-    const ext = extname(resolved);
-    if ((ext === '.js' || ext === '.cjs' || !CJSModule._extensions[ext]) &&
-        isAbsolute(resolved)) {
-      const { exportNames: reexportNames } = cjsPreparseModuleExports(resolved);
-      for (const name of reexportNames)
-        exportNames.add(name);
+    for (let i = 0; i < reexports.length; i++) {
+      const reexport = reexports[i];
+      let resolved;
+      try {
+        // TODO: this should be calling the `resolve` hook chain instead.
+        resolved = CJSModule._resolveFilename(reexport, module);
```

Comment:
It means the re-export of CJS loader will not always have the "correct" exported names – which would be good to fix, but I doubt it's going to block anyone from using the feature. I'm not going to fix it in this PR, feel free to do it (in this PR or in a follow up).

Reply:
🧐 this is the lexer step, what do you mean? The lexer step is relying on the CJS loader, which is unfortunate, but that needs to be the case for now (because it must be sync, so that would mean removing support for CJS in the loader thread, which I don't want to in this PR)

Reply:
Let’s back up a bit. I’m not really following what this comment is meaning or your first explanation of it. Like I don’t know what “the re-export of the CJS loader will not have the correct exported names” has to do with not calling the `resolve` hooks. Do you mind explaining this a bit more? And maybe the comment should be a bit more detailed.